src/mpi/coll/transports/generic/transport.c

```diff
@@ -0,0 +1,1243 @@
+                                       vtxp->nbargs.reduce_local.dt, vtxp->nbargs.reduce_local.op);
+
+                MPIR_Localcopy(vtxp->nbargs.reduce_local.inbuf,
+                               vtxp->nbargs.reduce_local.count,
+                               vtxp->nbargs.reduce_local.dt,
+                               vtxp->nbargs.reduce_local.inoutbuf,
+                               vtxp->nbargs.reduce_local.count, vtxp->nbargs.reduce_local.dt);
+                MPL_DBG_MSG_FMT(MPIR_DBG_COLL,VERBOSE,(MPL_DBG_FDEST,"  --> GENERIC transport (reduce local_R) complete\n"));
+            }
+
+            /* record vertex completion */
+            MPIR_COLL_GENERIC_record_vtx_completion(vtxp, sched);
+            break;
+        }
+    }
+
+    MPIR_FUNC_VERBOSE_EXIT(MPID_STATE_MPIR_COLL_GENERIC_ISSUE_VTX);
+}
+
+#undef FUNCNAME
+#define FUNCNAME MPIR_COLL_GENERIC_poke
+/* Transport function to make progress on the schedule */
+int MPIR_COLL_GENERIC_poke(MPIR_COLL_GENERIC_sched_t * sched)
+{
+    MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_MPIR_COLL_GENERIC_POKE);
+    MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPIR_COLL_GENERIC_POKE);
+
+    MPIR_COLL_GENERIC_vtx_t *vtxp;
+    int i;
+
+    /* If issued list is empty, go over the vertices and issue ready vertices
+     * Issued list should be empty only when the MPIR_COLL_GENERIC_poke function
+     * is called for the first time on the schedule */
+    if (sched->issued_head == NULL) {
+        MPL_DBG_MSG_FMT(MPIR_DBG_COLL,VERBOSE,(MPL_DBG_FDEST,"issued list is empty, issue ready vtcs\n"));
+        if (sched->total > 0 && sched->num_completed != sched->total) {
+            /* Go over all the vertices and issue ready vertices */
+            for (i = 0; i < sched->total; i++)
+                MPIR_COLL_GENERIC_issue_vtx(i, &sched->vtcs[i], sched);
+            MPL_DBG_MSG_FMT(MPIR_DBG_COLL,VERBOSE,(MPL_DBG_FDEST,"completed traversal of vtcs, sched->total: %d, sched->num_completed: %d\n",
+                     sched->total, sched->num_completed));
+            return 0;
+        } else
+            return 1;   /* schedule completed */
+    }
+
+    if (sched->total == sched->num_completed) { /* if done, return 1 */
+        return 1;
+    }
+
+    /* iterate over the issued vertex list, starting at the head */
+    sched->vtx_iter = sched->issued_head;
+    /* reset the issued vertex list, the list will be created again */
+    sched->issued_head = NULL;
+
+    /* Check for issued vertices that have been completed */
+    while (sched->vtx_iter != NULL) {
+        vtxp = sched->vtx_iter;
+        /* vtx_iter is updated to point to the next issued vertex
+         * that will be looked at after the current vertex */
+        sched->vtx_iter = sched->vtx_iter->next_issued;
+
+        if (vtxp->state == MPIR_COLL_GENERIC_STATE_ISSUED) {
+
+            MPI_Status status;
+            MPIR_Request *mpid_req0 = vtxp->mpid_req[0];
+            MPIR_Request *mpid_req1 = vtxp->mpid_req[1];
```

Comment:
Removing the weird `recv_reduce` operation would simplify all of this code.

Reply:
We need to keep that operation. I will see if we can simplify this code.